components/homme/test_execs/thetal_kokkos_ut/dirk_ut.cpp

```diff
@@ -0,0 +1,415 @@
+static void init (Random& r, HybridVCoord& h) {
+  h.random_init(r.random_device()());
+  const auto hyai = create_mirror_view(h.hybrid_ai); deep_copy(hyai, h.hybrid_ai);
+  const auto hybi = create_mirror_view(h.hybrid_bi); deep_copy(hybi, h.hybrid_bi);
+  const auto hyam = create_mirror_view(h.hybrid_am); deep_copy(hyam, h.hybrid_am);
+  const auto hybm = create_mirror_view(h.hybrid_bm); deep_copy(hybm, h.hybrid_bm);
+  init_dirk_f90(hyai.data(), hybi.data(), &hyam(0)[0], &hybm(0)[0], h.ps0);
+}
+
+static bool equal (const Real& a, const Real& b) {
+  if (a != b)
+    printf("equal: a,b = %23.16e %23.16e re = %23.16e\n",
+           a, b, std::abs((a-b)/a));
+  return a == b;
+}
+
+template <typename V>
+void fill (Random& r, const V& a,
+           typename std::enable_if<V::rank == 3>::type* = 0) {
+  const auto am = create_mirror_view(a);
+  deep_copy(am, a);
+  for (int i = 0; i < a.extent_int(0); ++i)
+    for (int j = 0; j < a.extent_int(1); ++j)
+      for (int k = 0; k < a.extent_int(2); ++k)
+        for (int s = 0; s < dfi::packn; ++s)
+          am(i,j,k)[s] = r.urrng(-1,1); 
+  deep_copy(a, am);
+}
+
+template <typename V>
+void fill (Random& r, const V& a,
+           typename std::enable_if<V::rank == 4>::type* = 0) {
+  const auto am = create_mirror_view(a);
+  deep_copy(am, a);
+  for (int i = 0; i < a.extent_int(0); ++i)
+    for (int j = 0; j < a.extent_int(1); ++j)
+      for (int k = 0; k < a.extent_int(2); ++k)
+        for (int l = 0; l < a.extent_int(3); ++l)
+          for (int s = 0; s < dfi::packn; ++s)
+            am(i,j,k,l)[s] = r.urrng(-1,1); 
+  deep_copy(a, am);
+}
+
+template <typename V>
+void fill_inc (Random& r, const int nlev, const Real top, const Real bottom,
+               const V& a, typename std::enable_if<V::rank == 3>::type* = 0) {
+  const auto am = create_mirror_view(a);
+  deep_copy(am, a);
+  for (int i = 0; i < a.extent_int(0); ++i)
+    for (int j = 0; j < a.extent_int(1); ++j) {
+      Real* const p = &am(i,j,0)[0];
+      Real sum = 0;
+      for (int k = 0; k < nlev; ++k) {
+        p[k] = r.urrng(); 
+        sum += p[k];
+      }
+      const Real f = (bottom - top)/sum;
+      for (int k = 0; k < nlev; ++k)
+        p[k] *= f;
+    }
+  deep_copy(a, am);
+}
+
+template <typename V>
+void fill_mid (Random& r, const int nlev, const Real top, const Real bottom,
+               const V& a, typename std::enable_if<V::rank == 3>::type* = 0) {
+  const auto am = create_mirror_view(a);
+  deep_copy(am, a);
+  fill_inc(r, nlev, top, bottom, am);
+  for (int i = 0; i < a.extent_int(0); ++i)
+    for (int j = 0; j < a.extent_int(1); ++j) {
+      Real* const p = &am(i,j,0)[0];
+      auto delta = p[0];
+      p[0] = top + 0.5*delta;
+      for (int k = 1; k < nlev; ++k) {
+        const auto tmp = p[k];
+        p[k] = p[k-1] + 0.5*(delta + p[k]);
+        delta = tmp;
+      }
+    }
+  deep_copy(a, am);
+}
+
+template <typename V>
+void require_all_equal (const int nlev, const V& a, const V& b,
```

Comment:
This could be migrated to `share/cxx/utilities/TestUtils.hpp`, and make all our test cpp files a bit shorter. I'll make a not to myself.

Reply:
Until I'm done with dirk_ut, I prefer you not make modifications to it.